File: NX4/UGOPEN/SampleNXOpenApplications/.NET/NXOpenExamples/EX_Curve_ProjCurves.vb
' ==================================================================================================
'
'            Copyright (c) 2004 UGS PLM Solutions.
'                   Unpublished - All rights reserved
'
'====================================================================================================
'File description:
'
'Test script EX_Curve_ProjCurves
'
'====================================================================================================
'   Date      Name                    Description of Change
'06-May-2004  Trupti Sahasrabudhe     Intial Creation
'19-May-2004  Alasdair Mackintosh     Use new TestSession class to obtain Session object
'16-May-2005  Trupti Sahasrabudhe     Convert C# Example to VB.Net
'$HISTORY$
'====================================================================================================
Option Strict Off
Imports System
Imports System.IO
Imports NXOpen
Imports NXOpen.Features
Imports NXOpen.Preferences
Imports NXOpen.UF

Module EX_Curve_ProjCurves

    Dim createMaster As Boolean = False ' set to True if creating a master file
    Dim testName As String = "EX_Curve_ProjCurves"

    Dim theSession As Session
    Dim theUfSession As UFSession

    Public Sub Main()
        theSession = Session.GetSession()
        theUfSession = UFSession.GetUFSession()

        Dim fs As New FileStream("EX_Curve_ProjCurves.log", FileMode.Create, FileAccess.Write)
        Dim w As New StreamWriter(fs) ' create a stream writer 
        w.Write("Log Entry : \r\n")
        w.WriteLine("--Log entry goes here--")
        w.Flush() ' update underlying file

        If File.Exists("EX_Curve_ProjCurves.prt") Then
            w.WriteLine("Remove EX_Curve_ProjCurves.prt file from <Project Folder>\\bin\\Debug !!")
            w.WriteLine("EX_Curve_ProjCurves.prt already exists. !!")
            w.Close()
            Return
        End If

        Dim nError As Integer = Execute(w, theUfSession)
        w.WriteLine("End of Log File")
        w.Close()
        
    End Sub

    Function Execute(ByVal stream As StreamWriter, ByVal theUfSession As NXOpen.UF.UFSession) As Integer

        Dim UFPart As NXOpen.Tag
        Dim part_name As String = "EX_Curve_ProjCurves"
        Dim units As Integer = 2
        Dim name As String
        Dim num_faces, num_proj_curves As Integer
        Dim origin() As Double = {0.0, 0.0, 0.0}
        Dim block_tag As NXOpen.Tag
        Dim curves_to_proj(1) As NXOpen.Tag
        Dim proj_curve_feature As NXOpen.Tag
        Dim proj_curves() As NXOpen.Tag
        Dim defining_feature, defining_target, defining_curve As NXOpen.Tag
        Dim edge_lens() As String = {"300.0", "25.0", "150.0"}
        Dim line_data As NXOpen.UF.UFCurve.Line = New NXOpen.UF.UFCurve.Line
        Dim proj_data As NXOpen.UF.UFCurve.Proj = New NXOpen.UF.UFCurve.Proj
        Dim face_list() As NXOpen.Tag

        Try
            theUfSession.Part.[New](part_name, units, UFPart)
            stream.WriteLine("New part : ", testName, " Created")

            theUfSession.Modl.CreateBlock1(FeatureSigns.Nullsign, origin, edge_lens, block_tag)
            stream.WriteLine("Successful:Modl.CreateBlock1")

            Dim array1() As Double = {25.0, 50.0, 25.0}
            line_data.start_point = array1

            Dim array2() As Double = {275.0, 50.0, 125.0}
            line_data.start_point = array2

            theUfSession.Curve.CreateLine(line_data, curves_to_proj(0))
            stream.WriteLine("Successful:Curve.CreateLine")
            'stream.WriteLine("curves to project tag = ", curves_to_proj(0))

            theUfSession.Modl.AskFeatFaces(block_tag, face_list)
            stream.WriteLine("Successful:Modl.AskFeatFaces")

            theUfSession.Modl.AskListCount(face_list, num_faces)
            stream.WriteLine("Successful:Modl.AskListCount")

            proj_data.proj_type = 3
            Dim array3() As Double = {0.0, 1.0, 0.0}
            proj_data.proj_vec = array3
            proj_data.multiplicity = 2

            theUfSession.Curve.CreateProjCurves(1, curves_to_proj, face_list.Length, face_list, 3, proj_data, proj_curve_feature)
            stream.WriteLine("Successful:Curve.CreateProjCurves")
            'stream.WriteLine("proj UFCurve feature tag = ", proj_curve_feature)

            theUfSession.Curve.AskProjCurves(proj_curve_feature, num_proj_curves, proj_curves)
            stream.WriteLine("Successful:Curve.AskProjCurves")

            For i As Integer = 0 To num_proj_curves - 1
                theUfSession.Curve.AskProjCurveParents(proj_curves(i), defining_feature, defining_target, defining_curve)
                stream.WriteLine("Successful:Curve.AskProjCurveParents")
                'stream.WriteLine("proj_curves ", i)
                'stream.WriteLine("  belongs to feature tag = ", defining_feature)
                'stream.WriteLine("  was projected onto tag = ", defining_target)
                'stream.WriteLine("  was generated by UFCurve tag = ", defining_curve)
            Next

            theUfSession.Part.Save()

        Catch ex As Exception
            stream.WriteLine("EXCEPTION: ", ex.Message)
            Return 1
        End Try

        Return 0

    End Function

End Module

File: NX1926_NX1930_NX1934_NX1938_NX1942_NX1946/UGOPEN/SampleNXOpenApplications/.NET/NXOpenExamples/EX_Curve_ProjCurves.cs
/*=============================================================================

                    Copyright (c) 2004 UGS PLM Solutions
                    Unpublished - All rights reserved

===============================================================================
File description: Example for Wrapped UFUNC (NXOpen.NET)
/*  */
using System;
using System.IO;
using NXOpen;
using NXOpen.UF;

namespace NetExample
{
    /// This example will create EX_Curve_ProjCurves.prt, EX_Curve_ProjCurves.log files 
    /// and demonstrates curve projection functionality usung theUfSession.Curve.CreateProjCurves().
    /// The program prints the report to the log file called "EX_Curve_ProjCurves.log".
    public class EX_Curve_ProjCurves
    {
        private static FileStream fs; 
        private static StreamWriter w;
        private static UFSession theUfSession;
        private static Session theSession;

       
            
        public int Execute()
        {
            Tag UFPart;
            string part_name = "EX_Curve_ProjCurves";
            int units =2; 
            string name;


            theUfSession.Part.New(part_name, units, out UFPart);
            theUfSession.Part.AskPartName(UFPart, out name);
            w.WriteLine("Loaded: " + name);
            
            int num_faces,i,num_proj_curves;
            double[] origin={0.0,0.0,0.0};
            Tag block_tag;
            Tag[] curves_to_proj= new Tag[1];
            Tag proj_curve_feature;
            Tag[] proj_curves;
            Tag defining_feature,defining_target,defining_curve;
            String[] edge_lens={"300.0","25.0","150.0"};
            UFCurve.Line line_data= new UFCurve.Line();
            UFCurve.Proj proj_data = new UFCurve.Proj();
            Tag[] face_list;
            theUfSession.Modl.CreateBlock1(FeatureSigns.Nullsign,origin,edge_lens,out block_tag);
            
            line_data.start_point = new double[3];
            line_data.start_point[0] =  25.0;
            line_data.start_point[1] =  50.0;
            line_data.start_point[2] =  25.0;
            line_data.end_point = new double[3]; 
            line_data.end_point[0]   = 275.0;
            line_data.end_point[1]   =  50.0;
            line_data.end_point[2]   = 125.0;
            
            theUfSession.Curve.CreateLine(ref line_data, out curves_to_proj[0]);
            Console.WriteLine("curves to project tag = {0}\n",curves_to_proj[0]);
            theUfSession.Modl.AskFeatFaces(block_tag,out face_list);
            theUfSession.Modl.AskListCount(face_list, out num_faces);
            
            proj_data.proj_type = 3;
            proj_data.proj_vec = new Double[3];
            proj_data.proj_vec[0] = 0.0;
            proj_data.proj_vec[1] = 1.0;
            proj_data.proj_vec[2] = 0.0;
            proj_data.multiplicity = 2;
            theUfSession.Curve.CreateProjCurves(1,curves_to_proj,face_list.Length,face_list,3,ref proj_data,out proj_curve_feature);
            w.WriteLine("proj UFCurve feature tag = {0}\n",proj_curve_feature);
            theUfSession.Curve.AskProjCurves(proj_curve_feature,out num_proj_curves,out proj_curves);
            for (i = 0;i < num_proj_curves;i++)
            {
                theUfSession.Curve.AskProjCurveParents(proj_curves[i],out defining_feature,out defining_target,out defining_curve);
                w.WriteLine("proj_curves[{0}]\n",i);
                w.WriteLine("  belongs to feature tag = {0}\n",defining_feature);
                w.WriteLine("  was projected onto tag = {0}\n",defining_target);
                w.WriteLine("  was generated by UFCurve tag = {0}\n", defining_curve);
            }

            theUfSession.Part.Save();

            return 0;
        }
        public static void Main(string[] args)
        {
            theSession=Session.GetSession();
            theUfSession= UFSession.GetUFSession();

            fs = new FileStream("EX_Curve_ProjCurves.log", FileMode.Create, FileAccess.Write);
            w = new StreamWriter(fs); // create a stream writer 
            w.Write("Log Entry : \r\n");
            w.WriteLine("--Log entry goes here--"); 
            w.Flush(); // update underlying file

            if ( File.Exists("EX_Curve_ProjCurves.prt") ) 
            {
                w.WriteLine("Remove EX_Curve_ProjCurves.prt file from <Project Folder>\\bin\\Debug !!");
                w.WriteLine("EX_Curve_ProjCurves.prt already exists. !!");
                w.Close();
                return;
            }

            try
            {
                EX_Curve_ProjCurves curveTest1 = new EX_Curve_ProjCurves();
                if (curveTest1.Execute()==0)
                {
                    w.WriteLine("Successful");
                }
                else
                {
                    w.WriteLine("Failed");
                }
            }
            catch(NXException e)
            {   
                w.WriteLine("Exception is: {0}", e.Message);
            }
            w.WriteLine("End of Log File");
            w.Close();
        }
        public static int GetUnloadOption(string dummy) 
        {
            return UFConstants.UF_UNLOAD_IMMEDIATELY;
        }
    }
}
